### PR TITLE
change p_keep autotuning behavior (fixes #380)

### LINKED ISSUE #380: bug in autotuning p_keep
in RISE.py line 119 the std is returning a single number. In line 120 the mean is taken which is a no-op for a single number. It doesn't really do what we thought it would do at the moment. It might still work as is though. We should add axis=n for the right axis when calling std.

### PR DESCRIPTION
**Pull request by cwmeijer**
Fixes #380 
This bug and its fix takes some knowledge of how RISE works to review.

We have 3 behaviors to talk about: The old actual behavior (that we considered bugged) which is the code of the main branch, the old intended behavior, and the new behavior as proposed in this PR.

### Purpose of the code
Note that the ultimate goal here is to pick the best p_keep. For this we calculate a score for every p_keep. 
We do this by masking instances with N random masks and make predictions for all. The result is a matrix of N by C (`number of classes`). We then calculate a score based on those and pick the p_keep with the highest score. This was the same for all 3 behaviors. The way the score was calculated was different.

#### Old intended behavior
The old intended behavior was as follows. For each of the classes we take the std over all N predictions. We then take the mean over all classes and end up with score that is a single scalar. 

#### Old actual behavior
For whatever reason, intentional or not, the old actual behavior was different. Of the `NxC` predictions we took the highest class of every N rows which resulted in a vector of length N. We then took the std of that, which is peculiar as the vector now contains predictions potentially belonging to varying classes. The result is a scalar, but we took the mean of that anyway, which is a no-op, indicating that we didn't really know what we were doing. The result was the score.

I didn't like the the old actual behavior because of mixing predictions from various classes and then doing statistics on that mix doesn't seem to make sense and I wouldn't know how to reason about it and pick a preferred p_keep based on that. I didn't like the old intended behavior either because it takes into account the stds of all classes. We are mostly interested in the top class or classes. In the case of a model with many classes (eg. ImageNet) the influence of the scores on irrelevant classes could be great and is unpredictable.

#### New behavior
Therefore, the new behavior that this PR proposes:
We take the std over all N rows and end up with a vector of length C, containing the std per class. We then take the max value of this, which we use as the score. 

Reasoning for the new behavior: Low scoring classes are bound to have low stds. Therefore the greatest std, belongs probably to the (or one of the) highest scoring classes. Alternatively, we could take the std of the highest scoring class instead for each of the N rows. This could be fine, but that would ignore all changes in probabilities, even of potentially important ones like other top classes.

Note that, because I changed the behavior, I also changed the tests.

### CODE REVIEW

File: dianna/methods/rise.py [hunk 1]
```diff
@@ -13,6 +13,17 @@ def _upscale(grid_i, up_size):
     return resize(grid_i, up_size, order=1, mode='reflect', anti_aliasing=False)
 
 
+def _predict_in_batches(masked, runner):
+    batch_size = 50
+    predictions = []
+    for i in range(0, len(masked), batch_size):
+        current_input = masked[i:i + batch_size]
```
[the remark this answers is not in the document]
Comment (cwmeijer) [reply dianna/methods/rise.py:20]: So, awktually... this already works as is. See:
image
While making a fix, I remembered a similar case this week while pair programming with Patrick. Apparently, slicing is always save in this respect that it doesn't generate any index out of bounds errors (see screenshot). I personally think this is unexpected behavior from Python (and you thought the same I suppose), so I'm going to apply one of your proposed fixes anyway.

**Comment by cwmeijer** [comment]
@loostrum , I addressed both of your comments. Please approve ;-)

File: dianna/methods/rise.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (loostrum) [reply dianna/methods/rise.py]: Huh I checked the range() function but not the indexing itself. Interesting that it does work! And I agree it's unexpected, so thanks for applying the fix.

**Review by loostrum — APPROVED** [review]

**Comment by loostrum** [comment]
Ah there's the signing commits error again. I applied signing to my commits by:
1. `git config --global user.signingkey='<public ssh key here that you use for auth with github>'`
2. `git config --global gpg.format=ssh`
3. `git config --global commit.gpgsign=true`

However I got some errors when trying to commit some things a few days later, so I had to disable the signing again... (you only need to set the config option of step 3 to false to do this)

**Pull request merged**